sctp/src/queue/pending_queue.rs

```diff
@@ -23,21 +33,52 @@ pub(crate) struct PendingQueue {
     unordered_is_selected: AtomicBool,
 }
 
+impl Default for PendingQueue {
+    fn default() -> Self {
+        PendingQueue::new()
+    }
+}
+
+// Some tests push a lot of data before starting to process any data...
+#[cfg(test)]
+const QUEUE_BYTES_LIMIT: usize = 128 * 1024 * 1024;
+#[cfg(not(test))]
+const QUEUE_BYTES_LIMIT: usize = 128 * 1024;
+
+const QUEUE_APPEND_LARGE: usize = (QUEUE_BYTES_LIMIT * 2) / 3;
+
 impl PendingQueue {
     pub(crate) fn new() -> Self {
-        PendingQueue::default()
+        Self {
+            semaphore_lock: Mutex::default(),
+            semaphore: Semaphore::new(QUEUE_BYTES_LIMIT),
+            unordered_queue: Default::default(),
+            ordered_queue: Default::default(),
+            queue_len: Default::default(),
+            n_bytes: Default::default(),
+            selected: Default::default(),
+            unordered_is_selected: Default::default(),
+        }
     }
 
     /// Appends a chunk to the back of the pending queue.
-    pub(crate) fn push(&self, c: ChunkPayloadData) {
+    pub(crate) async fn push(&self, c: ChunkPayloadData) {
         let user_data_len = c.user_data.len();
 
-        if c.unordered {
-            let mut unordered_queue = self.unordered_queue.write();
-            unordered_queue.push_back(c);
-        } else {
-            let mut ordered_queue = self.ordered_queue.write();
-            ordered_queue.push_back(c);
+        {
+            let sem_lock = self.semaphore_lock.lock().await;
+            let permits = self.semaphore.acquire_many(user_data_len as u32).await;
+            // unwrap ok because we never close the semaphore unless we have dropped self
+            permits.unwrap().forget();
+
+            if c.unordered {
+                let mut unordered_queue = self.unordered_queue.write();
+                unordered_queue.push_back(c);
+            } else {
+                let mut ordered_queue = self.ordered_queue.write();
+                ordered_queue.push_back(c);
+            }
+            drop(sem_lock);
```

Comment:
are you sure we need to drop lock manually? why it's not dropped automatically?

Reply:
Nope it's not required, but for functions where the Lock is not needed the whole time I like to do it like this. I know this would be dropped and unlocked at the end of the block anyways but I think it communicates intent better.

If you don't like it, it can be removed.

Reply:
Especially since the construct with the sem_lock isn't required by any language level checks to actually push to the queue I think it is good to make sure the lock actually lives as long, but not longer than needed.